### DISCUSSION TITLE
Ruminating / trying to get a handle on things  =)

### DISCUSSION THREAD

**Discussion by metaleap** [godot-go/godot-go#102] [General]
As I was lately getting into Godot, discovering the GDExtension stuff had me hooked, the idea of non-script game-dev with that impressive engine really appealed to me — except C++ wouldn't.

Now I've been pondering between Golang, Zig or plain C, all of which I'd pick over C++ any day.

Well, with Zig and C the story would be to link the `.h` included-with-or-freshly-regenerated-with https://github.com/gilzoide/gdextension-lite (which uses the `.json` dumped by one's currently-installed `godot --dump-extension-api`) — kinda not too daunting at first glance. Now Zig would be even sweeter than C if it had a stable 1.x (been burned before with ever-more breaking-changes), but seems like a long way off even after so many years. Fair enough, bummer tho  =)

**Either way, Golang would still appeal a _bit_ more to me here** than those 2... the GC is tunable enough and in a game one's allocations can be somewhat batched/front-loaded, pools etc, no big deal.

Well for me that raises the question how this project is sorta approached by you in design terms. Because from the above plain-C `.h` generator on one hand and the long-standing existence of CGO on another, I would have thought that nowadays there's tooling to generate a mostly-CGO pkg from any C API. (Wild assumption about the-unknown here, as I avoided CGO all these years.) Yet clearly this repo shows much hand-crafted development and a [roadmap/checklist](https://github.com/godot-go/godot-go/blob/main/docs/overview.md#container-built-in-types) too.

I guess all that is for conversions to/from unsafe ptrs / struct-layout compat / go-native feeling / C-string-Go-string-and-GD-String conversions.

How do you see the timeframe / complexity / daunting-ness of the remaining container types currently marked "not yet implemented"? Is every one of them a totally uniquely different beast than all the others? I guess the nums are simpler than the various others... have a clear idea how it should principally be done, or stuck on something? Will see myself how I grok this repo shortly, but just rambling a little.. cheers, thanks for anyone's thoughts here  =)

AND — how is your actual underlying C link-up, surely with CGO it's not godot-cpp which is C++, but some `.h` or `.a` or `.o`?

**Comment by metaleap** [comment]
Ah well, I spat on my sensibilities / prejudices and did get into C++. It's actually grooving plenty fine.  =)

**Reply by pcting** [reply]
sorry. i've been busy with the holidays and work. hopefully i can resolve some of the questions:

i started this godot-go journey to understand go better and to find a way to contribute back to the godot community.

cgo tooling (or the lack of tooling) has made building these bindings very difficult to put together.  i've learned a lot and would caution anyone to not build a go as c shared lib if there are other options. high level problems:
* stacktraces are lost when jumping between C code and go code making it very difficult to track down crashes. this is the reason for all the `printStacktrace` calls in the C wrapper code.
* go doesn't fully unload cleanly if you plan to do this multiple times in a C process. open ticket: https://github.com/golang/go/issues/11100
* GDExtension exposes opaque data blobs and pointers to native godot data types and class instances respectively. Interop / marshalling / unmarshalling with native go types requires hand-crafting, which is why there's so much meta-programming / codegen in the bindings (https://github.com/godot-go/godot-go/tree/main/cmd/generate).  the `extension_api.json` does contain enough data to get byte offsets for the data types, so it should be possible to manipualte those without jumping into godot code.
* running as a c-shared lib, i haven't found a way to debug go code with `dlv`. i've been debugging with `gdb`. this is fine and all, but inspecting go data types is more difficult as you will have to interpret the go internal memory layout to understand what you are looking at.

implementing the remaining missing array-like data types shouldn't be hard; i just haven't had the time to focus on it. once one of them gets implemented, the rest should be pretty quick.

the linking is all done with the standard cgo tooling. if you go through the "Getting Started" section, you'll eventually end up with a `test/demo/lib/libgodotgo-test-linux-amd64.so` file that godot will load. this is done by specifying that go build should output a `c-shared` lib (https://github.com/godot-go/godot-go/blob/48ae418ad7b7f207ef3ab06ccd2c51f51359be19/Makefile#L55C112-L55C131).

one thing i haven't really focused on is performance. my immediate goal is to get the godot-go binding feature complete before focusing on optimizations. memory layout of the non-trivial types.

hope this helps.

Cheers!
